### PR TITLE
Add a proper ESLint config

### PR DESCRIPTION
**Pull request by vostrnad**
Let's break GitHub with a few hundred ESLint annotations.

Please check the generated ESLint warnings. If you see any warning that you would not like bothering you, we shall discuss it.

### CODE REVIEW

File: app/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        "arrow-parens": "error",
+        "arrow-spacing": "error",
+        "no-confusing-arrow": "error",
+        "no-duplicate-imports": "error",
+        "no-useless-rename": "error",
+        "object-shorthand": "error",
+        "prefer-const": "error",
+        "prefer-rest-params": "error",
+        "prefer-spread": "error",
+        "prefer-template": "error",
```
Comment (Maelstromeous) [inline app/.eslintrc.js:307]: Will need this for some string concatination, bit overkill to use templates for everything. Suggest turning to warning.

File: app/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+                "format": ["UPPER_CASE"]
+            }
+        ],
+        "@typescript-eslint/no-base-to-string": "error",
+        "@typescript-eslint/no-confusing-non-null-assertion": "error",
+        "@typescript-eslint/no-dynamic-delete": "error",
+        "@typescript-eslint/no-invalid-void-type": "error",
+        "@typescript-eslint/no-parameter-properties": "error",
+        "@typescript-eslint/no-require-imports": "error",
+        "@typescript-eslint/no-throw-literal": "error",
```
Comment (Maelstromeous) [inline app/.eslintrc.js:134]: This breaks our way of throwing exceptions, could you offer a better way of doing it or remove this?

E.g. ContinentLockEvent, various throws there.

File: app/.eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+            },
+            {
+                "selector": "property",
+                "modifiers": ["static", "readonly"],
+                "types": ["boolean", "string", "number"],
+                "format": ["UPPER_CASE"]
+            },
+            {
+                "selector": "enumMember",
+                "format": ["UPPER_CASE"]
```
Comment (Maelstromeous) [inline app/.eslintrc.js:125]: On the fence about ENUMs being uppercase, this looks fucky when it comes to say the faction names.

File: app/.eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    ],
+    "rules": {
+        // @typescript-eslint rules
+        "@typescript-eslint/array-type": [
+            "error",
+            {
+                "default": "array-simple"
+            }
+        ],
+        "@typescript-eslint/consistent-type-definitions": [
```
Comment (Maelstromeous) [inline app/.eslintrc.js:26]: This needs further explanation

File: app/.eslintrc.js (hunk 1 above)
Comment (vostrnad) [reply app/.eslintrc.js:307]: Templates stop feeling like overkill very soon once you start using them as the standard way. But I have nothing against switching it to warning.

File: app/.eslintrc.js (hunk 2 above)
Comment (vostrnad) [reply app/.eslintrc.js:134]: Look into ApplicationException.ts. I made a small adjustment that resolves this.

File: app/.eslintrc.js (hunk 4 above)
Comment (vostrnad) [reply app/.eslintrc.js:26]: This means preferring `interface` over `type` when declaring definitions for objects. Apart from a few exceptions we were already doing it like this, so it's good to keep it consistent.

File: app/.eslintrc.js (hunk 3 above)
Comment (vostrnad) [reply app/.eslintrc.js:125]: Enum members have conventionally always been in UPPER_CASE. Check out the Java docs for example: https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html

File: app/.eslintrc.js (hunk 4 above)
Comment (Maelstromeous) [reply app/.eslintrc.js:26]: Well I see this as requiring the class to implement an interface, but the cases where it's come up, it makes sense to be a type?

File: app/.eslintrc.js (hunk 3 above)
Comment (Maelstromeous) [reply app/.eslintrc.js:125]: Alright, I'll let this one go as PHP Constants are uppercase as well.

File: app/.eslintrc.js (hunk 1 above)
Comment (Maelstromeous) [reply app/.eslintrc.js:307]: Yeah lets go ahead with warning for now, if they're proving bothersome we can make it error.

File: app/.eslintrc.js (hunk 2 above)
Comment (Maelstromeous) [reply app/.eslintrc.js:134]: Ah gotcha

File: app/.eslintrc.js (hunk 4 above)
Comment (vostrnad) [reply app/.eslintrc.js:26]: `type` is used for all kinds of type definitions, `interface` only for object definitions. Here is an article if you want more: https://pawelgrzybek.com/typescript-interface-vs-type/

Comment (Maelstromeous) [reply app/.eslintrc.js:26]: Upon reviewing the article, I've been swayed. Resolving.

**Pull request merged**